### PR TITLE
Implement Lab3 - Letter Frequency

### LINKED ISSUE #262: Lab 3 - Letter Frequency
Create an executable go program in directory 03_letters/brvishnu92
Write a function that returns a mapping of each letter to its frequency
Write a function that returns a sorted slice of strings with elements "{key}:{val}"
Write test cases covering all function runs.

### PR DESCRIPTION
**Pull request by brvishnu92**
Fixes #262

Review of colleague's PR #343

#### Changes proposed in this PR:

-  Implement Lab3 LetterFrequency.
 - Test Lab3.

### CODE REVIEW

File: 03_letters/brvishnu92/main_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+func TestMain(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+	expected := `a:2
+b:1
+`
+	actual := buf.String()
+	assert.Equalf(t, expected,actual,  "got %v want %v", actual,expected)
```
Comment (golangcibot) [inline 03_letters/brvishnu92/main_test.go]: File is not `goimports`-ed (from `goimports`)

File: 03_letters/brvishnu92/main.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+func main() {
+	fmt.Fprintln(out, strings.Join(sortLetters(letters("aba")), "\n"))
+}
+
+func letters(s string) map[rune]int {
+
+	x := make(map[rune]int)
+	for _, char := range s {
+			x[char]++
```
Comment (golangcibot) [inline 03_letters/brvishnu92/main.go]: File is not `goimports`-ed (from `goimports`)

File: 03_letters/brvishnu92/main.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, strings.Join(sortLetters(letters("aba")), "\n"))
+}
+
+func letters(s string) map[rune]int {
+
+	x := make(map[rune]int)
```
Comment (CameronSchafer) [inline 03_letters/brvishnu92/main.go:18]: Is this space needed here?

File: 03_letters/brvishnu92/main.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	for _, char := range s {
+		x[char]++
+	}
+	return x
+}
+
+func sortLetters(m map[rune]int) []string {
+
+	keys := []int{}
```
Comment (CameronSchafer) [inline 03_letters/brvishnu92/main.go]: Is this space needed here?

**Review by CameronSchafer — CHANGES_REQUESTED** [review]
This code LGTM
Just a couple of nitpicks about extra spaces.

**Review by CameronSchafer — DISMISSED** [review]

File: 03_letters/brvishnu92/main.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintln(out, strings.Join(sortLetters(letters("aba")), "\n"))
+}
+
+func letters(s string) map[rune]int {
+	x := make(map[rune]int)
+	for _, char := range s {
```
Comment (camscale) [inline 03_letters/brvishnu92/main.go:19]: FYI: Go uses the term "rune" instead of "char" to avoid the issue of people thinking it might be an 8-bit value. No need to change this now, but a better name would have been just `r` (for rune). `for _, r := range s` is an obvious Go expression (iterating the runes in a string)

File: 03_letters/brvishnu92/main_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	actual := buf.String()
+	assert.Equalf(t, expected, actual, "got %v want %v", actual, expected)
+}
+
+func TestLetters(t *testing.T) {
+	testCases := map[string]struct {
+		input    string
+		expected map[rune]int
+	}{
+		"test1":   {input: "blahbla", expected: map[rune]int{97: 2, 98: 2, 104: 1, 108: 2}},
```
Comment (camscale) [inline 03_letters/brvishnu92/main_test.go]: These tests would all be a lot clearer if you used the rune as the key and not its integer value: `map[rune]int{'a': 2, 'b': 2, 'h': 1, 'l': 2}`

File: 03_letters/brvishnu92/main_test.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		"test2":   {input: "smarty", expected: map[rune]int{97: 1, 109: 1, 114: 1, 115: 1, 116: 1, 121: 1}},
+		"empty":   {input: "", expected: map[rune]int{}},
+		"sameval": {input: "aaaa", expected: map[rune]int{97: 4}},
+		"oneval":  {input: "a", expected: map[rune]int{97: 1}},
+	}
```
Comment (camscale) [inline 03_letters/brvishnu92/main_test.go:31]: Please add a test with non-ascii unicode runes in the string - this makes sure your unit under test is treating the elements of a string as runes and not bytes (which can be a common mistake for people coming from other languages).

File: 03_letters/brvishnu92/main_test.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+			assert.Equalf(t, test.expected, result, "got %v want %v", result, test.expected)
+		})
+	}
+}
+
+func TestSortLetters(t *testing.T) {
+	testCases := map[string]struct {
+		expected []string
+		input    map[rune]int
+	}{
```
Comment (camscale) [inline 03_letters/brvishnu92/main_test.go:45]: Likewise, please use rune literals as the map keys instead of integers. It is more readable.

File: 03_letters/brvishnu92/main_test.go (hunk 6 above)
Comment (camscale) [inline 03_letters/brvishnu92/main_test.go]: You should give your tests meaningful names, and only add tests that test additional cases. For "test1" and "test2", they are really testing the same thing, so you only need one of them. Additional unnecessary tests just add a maintenance burden for no additional gain.

If you can come up with meaningful names for your tests, you can often see that two tests would have the same name, so there's no point having both.

File: 03_letters/brvishnu92/main_test.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		"test1":   {expected: []string{"a:2", "b:2", "h:1", "l:2"}, input: map[rune]int{97: 2, 98: 2, 104: 1, 108: 2}},
+		"empty":   {expected: []string{}, input: map[rune]int{}},
+		"sameval": {expected: []string{"a:4"}, input: map[rune]int{97: 4}},
+		"oneval":  {expected: []string{"a:1"}, input: map[rune]int{97: 1}},
```
Comment (camscale) [inline 03_letters/brvishnu92/main_test.go]: "sameval" and "oneval" are really the same test in that the code does the same thing in both cases. You only need one of them.

**Review by camscale — CHANGES_REQUESTED** [review]
The Go code is good, but I think the tests need a little work. Some comments are general software engineering oriented, but there are some Go-specific comments there that need to be addressed for this Go course, so you may was well address them all.

File: 03_letters/brvishnu92/main_test.go (hunk 6 above)
Comment (brvishnu92) [reply 03_letters/brvishnu92/main_test.go]: Sure!! :-D Done.

File: 03_letters/brvishnu92/main_test.go (hunk 7 above)
Comment (brvishnu92) [reply 03_letters/brvishnu92/main_test.go:31]: Done. Added special char and emojis

File: 03_letters/brvishnu92/main_test.go (hunk 8 above)
Comment (brvishnu92) [reply 03_letters/brvishnu92/main_test.go:45]: Done.

File: 03_letters/brvishnu92/main_test.go (hunk 9 above)
Comment (brvishnu92) [reply 03_letters/brvishnu92/main_test.go]: Yeah, I agree. Have removed the test case.

File: 03_letters/brvishnu92/main_test.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func TestSortLetters(t *testing.T) {
+	testCases := map[string]struct {
+		expected []string
+		input    map[rune]int
+	}{
+		"normal":  {expected: []string{"h:1", "k:2", "l:2", "z:2"}, input: map[rune]int{'z': 2, 'k': 2, 'h': 1, 'l': 2}},
+		"empty":   {expected: []string{}, input: map[rune]int{}},
+		"special": {expected: []string{"^:4"}, input: map[rune]int{'^': 4}},
```
Comment (rokane) [inline 03_letters/brvishnu92/main_test.go]: I am a little confused at what this is testing for. From the sound of the description it appears as though you are looking at the sorting of special runes, however there is really only one rune in the map so it will only ever be sorted one way, doesn't matter if it is a special or normal char. Maybe add a couple more special runes and test the sorting of such, or change this to be named "single" if that was what you were hoping to test.

**Review by rokane — CHANGES_REQUESTED** [review]
Looking really good. Just have one minor suggestion with the `TestSortLetters` test case.

File: 03_letters/brvishnu92/main_test.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func TestSortLetters(t *testing.T) {
+	testCases := map[string]struct {
+		expected []string
+		input    map[rune]int
+	}{
+		"normal":  {expected: []string{"h:1", "k:2", "l:2", "z:3"}, input: map[rune]int{'z': 3, 'k': 2, 'h': 1, 'l': 2}},
+		"empty":   {expected: []string{}, input: map[rune]int{}},
+		"special": {expected: []string{"*:2","@:3","^:4"}, input: map[rune]int{'^': 4, '*':2, '@':3}},
```
Comment (golangcibot) [inline 03_letters/brvishnu92/main_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
		"special": {expected: []string{"*:2", "@:3", "^:4"}, input: map[rune]int{'^': 4, '*': 2, '@': 3}},
```

File: 03_letters/brvishnu92/main_test.go (hunk 10 above)
Comment (brvishnu92) [reply 03_letters/brvishnu92/main_test.go]: Done. Lol I was trying to show special characters was being processed but forgot about the function sorting it :-)

**Comment by brvishnu92** [comment]
@rokane Can you please have a look at the updated code when you are free? Thanks a ton :-)

File: 03_letters/brvishnu92/main_test.go (hunk 6 above)
Comment (camscale) [reply 03_letters/brvishnu92/main_test.go]: I've been waiting for you to push these changes before re-reviewing, but perhaps you just forgot.

You said "Done", but I don't see that it has been done. Please double-check your response to comments before asking for re-review. Thanks

**Review by camscale — DISMISSED** [review]
Changes I asked for last time still have not been made. Please go through all the comments and check your code. Thanks.

**Comment by brvishnu92** [comment]
@camh-anz  Sorry, I had implemented the changes but lost them while rebasing. I tried to do the same now, but again i could see the old file reflecting while rebasing after working on it. So i overwrote the file again and pushed.

**Comment by brvishnu92** [comment]
Maybe because I work off two laptops.. I do “git pull” but should I be doing “git fetch” as well ...

**Review by rokane — APPROVED** [review]
LGTM 👍
